blockchain/sequencelock.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2017 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package blockchain
+
+import (
+	"fmt"
+
+	"github.com/decred/dcrd/blockchain/stake"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrutil"
+)
+
+// SequenceLock represents the minimum timestamp and minimum block height after
+// which a transaction can be included into a block while satisfying the
+// relative lock times of all of its input sequence numbers.  It is calculated
+// via the CalcSequenceLock function.  Each field may be -1 if none of the input
+// sequence numbers require a specific relative lock time for the respective
+// type.  Since all valid heights and times are larger than -1, this implies
+// that it will not prevent a transaction from being included due to the
+// sequence lock, which is the desired behavior.
+type SequenceLock struct {
+	MinHeight int64
+	MinTime   int64
+}
+
+// isStakeBaseTx determines whether or not a transaction is a stakebase (also
+// known as a vote).  A stakebase is a special transaction created by the
+// proof-of-stake system that creates subsidy.  This function simply delegates
+// to the IsSSGen function in the stake package and exists to make calling code
+// that does not care about the specific reason the transaction is not a
+// stakebase, rather only if it is one or not.
+func isStakeBaseTx(tx *wire.MsgTx) bool {
+	isStakeBase, _ := stake.IsSSGen(tx)
+	return isStakeBase
+}
+
+// calcSequenceLock computes the relative lock times for the passed transaction
+// from the point of view of the block node passed in as the first argument.
+//
+// See the CalcSequenceLock comments for more details.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcSequenceLock(node *blockNode, tx *dcrutil.Tx, view *UtxoViewpoint, isActive bool) (*SequenceLock, error) {
```

Comment:
should this always be the best node?  seems to be called only with it, and would make the code clearer if that was explicitly spelled out either in the function comment or the variable name.

Reply:
question arises because it uses that node's height + 1 to calculate the mempool height

Reply:
It has to be an arbitrary node because it is called for reorganizations once the consensus code that calls it based on the result of the required consensus vote (PR #855).